### PR TITLE
feature: changeable keybindings

### PR DESCRIPTION
**Pull request by littlemyx**
Adding an ability to change keybindings
For #168

### CODE REVIEW

File: src/controls/Keyboard.ts [hunk 1]
```diff
@@ -43,27 +34,7 @@ function useKeys(keyConfig: KeyConfig[]) {
...
-    { keys: ['m', 'M'], fn: () => set((state) => ({ map: !state.map })), up: false },
-    { keys: ['p', 'P'], fn: () => set((state) => ({ help: false, pickcolor: !state.pickcolor, leaderboard: false })), up: false },
-    { keys: ['u', 'U'], fn: () => set((state) => ({ sound: !state.sound })), up: false },
-    {
-      keys: ['c', 'C'],
-      fn: () => set((state) => ({ camera: cameras[(cameras.indexOf(state.camera) + 1) % cameras.length] })),
-      up: false,
-    },
-  ])
+  const [keyboardBindings] = useStore((state) => [state.keyboardBindings])
```
Comment (njm222) [inline src/controls/Keyboard.ts:37]:
```suggestion
  const keyboardBindings = useStore((state) => state.keyboardBindings)
```

File: src/controls/Keyboard.ts [hunk 2]
```diff
@@ -1,34 +1,25 @@
...
-  pressed?: boolean
-}
+import { useStore } from '../store'
+import type { KeyConfig, KeyMap } from '../store'
 
 function useKeys(keyConfig: KeyConfig[]) {
   useEffect(() => {
-    const keyMap = keyConfig.reduce<{ [key: string]: KeyMap }>((out, { keys, fn, up = true }) => {
-      keys && keys.forEach((key) => (out[key] = { fn, pressed: false, up }))
+    const codeMap = keyConfig.reduce<{ [key: string]: KeyMap }>((out, { keys, fn, up = true }) => {
```
Comment (njm222) [inline src/controls/Keyboard.ts:7]: why did you rename `keyMap` and `key`?

File: src/store.ts [hunk 3]
```diff
@@ -106,6 +124,28 @@ export interface IState extends BaseState {
...
+function deduplicateKeys(newKey: Key, keysList: KeyConfig[]): KeyConfig[] {
+  const deduplicatedList = keysList.map((key) => {
+    const index = key.keys.findIndex((keyCode) => keyCode.name === newKey.name)
+    if (index !== -1) {
+      key.keys.splice(index, 1)
+    }
+
+    return key
+  })
+  return deduplicatedList
```
Comment (njm222) [inline src/store.ts:139]:
```suggestion
  return keysList.map((key) => {
    const index = key.keys.findIndex((keyCode) => keyCode.name === newKey.name)
    if (index !== -1) {
      key.keys.splice(index, 1)
    }

    return key
  })
```

File: src/store.ts [hunk 4]
```diff
@@ -138,8 +178,109 @@ const useStoreImpl = create<IState>((set: SetState<IState>, get: GetState<IState
         return { ...state, finished: 0, start: 0 }
       })
     },
+    addKeyBinding: (action: string, newKey: Key) => {
+      set((state) => {
+        const index = state.keyboardBindings.findIndex(({ action: stateAction }) => action === stateAction)
+
+        const keyboardBindingsCopy = deduplicateKeys(newKey, state.keyboardBindings)
+
+        keyboardBindingsCopy[index].keys.push(newKey)
```
Comment (njm222) [inline src/store.ts:187]: should handle null `index`

File: src/store.ts [hunk 5]
```diff
@@ -138,8 +178,109 @@ const useStoreImpl = create<IState>((set: SetState<IState>, get: GetState<IState
...
+    },
+    removeKeyBinding: (action: string, name: string) => {
+      set((state) => {
+        const index = state.keyboardBindings.findIndex(({ action: stateAction }) => action === stateAction)
+
+        const keyIndex = state.keyboardBindings[index].keys.findIndex(({ name: stateName }) => name === stateName)
+
+        const keyboardBindingsCopy = [...state.keyboardBindings]
+
+        keyboardBindingsCopy[index].keys.splice(keyIndex, 1)
```
Comment (njm222) [inline src/store.ts:202]: should handle null `index` & `keyIndex`

File: src/store.ts [hunk 6]
```diff
@@ -138,8 +178,109 @@ const useStoreImpl = create<IState>((set: SetState<IState>, get: GetState<IState
...
+
+        const keyBindingsWithError = checkKeybindings(keyboardBindingsCopy)
+
+        return { ...state, keyboardBindings: keyboardBindingsCopy, keyBindingsWithError }
+      })
+    },
   }
 
+  const keyboardBindings: KeyConfig[] = [
```
Comment (njm222) [inline src/store.ts:211]:
```suggestion
  const defaultKeyboardBindings: KeyConfig[] = [
```

File: src/styles.css [hunk 7]
```diff
@@ -440,6 +466,20 @@ button {
   animation: hideSoundIcon 3s forwards;
 }
 
+.key-input {
```
Comment (njm222) [inline src/styles.css:469]: maybe create a wrapper to handle the absolute positioning and then flex for the child

File: src/ui/Help.tsx [hunk 8]
```diff
@@ -40,21 +26,171 @@ export function Help(): JSX.Element {
...
+  const [isHovered, setIsHovered] = useState(false)
+  return (
+    <div
+      className={`popup-item${isHovered ? ' hovered' : ''}${hasError ? ' with-error' : ''}`}
+      onMouseOver={() => {
+        setIsHovered(true)
+      }}
+      onMouseLeave={() => {
+        setIsHovered(false)
+      }}
```
Comment (njm222) [inline src/ui/Help.tsx:65]: use `:hover` instead of using state to handle this.
https://developer.mozilla.org/en-US/docs/Web/CSS/:hover

File: src/ui/Help.tsx [hunk 9]
```diff
@@ -40,21 +26,171 @@ export function Help(): JSX.Element {
...
+      window.removeEventListener('keydown', onSelect)
+    }
+  }, [onSelect])
+
+  return <div className="key-input">Press new key</div>
+}
+
+type KeyProps = HTMLAttributes<HTMLDivElement>
+
+export function Keys(props: KeyProps): JSX.Element {
```
Comment (njm222) [inline src/ui/Help.tsx:171]: extract this to its own file in /ui

**Review by njm222 — CHANGES_REQUESTED** [review]
Thanks a lot for this awesome feature! Great work so far, just some minor changes.

File: src/controls/Keyboard.ts (hunk 1 above)
Comment (littlemyx) [reply src/controls/Keyboard.ts:37]: The problem was in the inappropriate way of how I was creating deduplicated list of keys (traditional javascriptish reference-mistake 😞  ), now I've realised it and fixed, so now it is really no need in array here.

File: src/controls/Keyboard.ts (hunk 2 above)
Comment (littlemyx) [reply src/controls/Keyboard.ts:7]: At first there was another idea of how I was going to manage keys. Returned the old name

File: src/ui/Help.tsx (hunk 8 above)
Comment (littlemyx) [reply src/ui/Help.tsx:65]: with css-only approach it is impossible to manage the presence of the button in DOM (only being able to hide it visually)

File: src/store.ts (hunk 3 above)
Comment (littlemyx) [reply src/store.ts]: done

File: src/store.ts (hunk 4 above)
Comment (littlemyx) [reply src/store.ts]: done

File: src/store.ts (hunk 5 above)
Comment (littlemyx) [reply src/store.ts]: done

File: src/store.ts (hunk 6 above)
Comment (littlemyx) [reply src/store.ts]: done

File: src/styles.css (hunk 7 above)
Comment (littlemyx) [reply src/styles.css]: done

File: src/ui/Help.tsx (hunk 9 above)
Comment (littlemyx) [reply src/ui/Help.tsx]: done

**Comment by littlemyx** [comment]
Sorry for this long delay, as you may heard, one idiot-president had started a war against our beloved neighbours - Ukraine  (sad but true - I'm from russia 😞 ), so I had to leave my country urgently just not to support the government and wander a bit in search of new home.

File: src/store.ts [hunk 10]
```diff
@@ -106,6 +124,20 @@ export interface IState extends BaseState {
   vehicleConfig: VehicleConfig
   wheelInfo: WheelInfo
   wheels: [RefObject<Object3D>, RefObject<Object3D>, RefObject<Object3D>, RefObject<Object3D>]
+  keyInput: string | null
+}
+
+function deduplicateKeys(newKey: Key, keysList: KeyConfig[]): KeyConfig[] {
+  return keysList.map((key) => ({ ...key, keys: [...key.keys.filter((keyCode) => keyCode.name !== newKey.name)] }))
```
Comment (njm222) [inline src/store.ts:131]:
```suggestion
  return keysList.map((key) => ({ ...key, keys: key.keys.filter((keyCode) => keyCode.name !== newKey.name) }))
```

File: src/store.ts [hunk 11]
```diff
@@ -138,8 +170,113 @@ const useStoreImpl = create<IState>((set: SetState<IState>, get: GetState<IState
         return { ...state, finished: 0, start: 0 }
       })
     },
+    addKeyBinding: (action: string, newKey: Key) => {
+      set((state) => {
+        const index = state.keyboardBindings.findIndex(({ action: stateAction }) => action === stateAction)
+
+        const keyboardBindingsCopy = deduplicateKeys(newKey, state.keyboardBindings)
+
+        if (index !== -1) {
```
Comment (njm222) [inline src/store.ts:179]: make this:
```suggestion
        if (index < 0) {
```
move to the top and return and then the rest will run if it is a valid index

File: src/store.ts [hunk 12]
```diff
@@ -138,8 +170,113 @@ const useStoreImpl = create<IState>((set: SetState<IState>, get: GetState<IState
...
+    },
+    removeKeyBinding: (action: string, name: string) => {
+      set((state) => {
+        const index = state.keyboardBindings.findIndex(({ action: stateAction }) => action === stateAction)
+
+        const keyIndex = state.keyboardBindings[index].keys.findIndex(({ name: stateName }) => name === stateName)
+
+        const keyboardBindingsCopy = [...state.keyboardBindings]
+
+        if (index !== -1) {
```
Comment (njm222) [inline src/store.ts:196]: same

File: src/ui/Help.tsx (hunk 8 above)
Comment (njm222) [reply src/ui/Help.tsx]: I'm pretty sure there are a few ways of doing this.
what about using `opacity: 0`?

**Review by njm222 — CHANGES_REQUESTED** [review]
Almost there!

No worries, yes I heard ( even though I live under a rock ) I hope you are staying safe out there and 🤞 you find a new home soon! :)

File: src/store.ts (hunk 10 above)
Comment (littlemyx) [reply src/store.ts:131]: done

File: src/store.ts (hunk 11 above)
Comment (littlemyx) [reply src/store.ts:179]: done

File: src/store.ts (hunk 12 above)
Comment (littlemyx) [reply src/store.ts:196]: done

File: src/store.ts (hunk 10 above)
Comment (littlemyx) [reply src/store.ts]: done

File: src/ui/Help.tsx (hunk 8 above)
Comment (littlemyx) [reply src/ui/Help.tsx]: holding this elements in the DOM will also hold them in vDOM what is non-optimal due to the fact that only one line can be hovered in one moment (but yes, it is kind of trade of between holding nodes in memory and holding callbacks end event listeners in it)

Comment (njm222) [reply src/ui/Help.tsx]: yes I understand but the css `:hover` is soo much faster and less buggy. Using state will cause an unnecessary rerender.

Comment (littlemyx) [reply src/ui/Help.tsx]: see your point and accepting it

File: src/ui/Keys.tsx [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+            key={name}
+            onClick={() => {
+              onRemove(action, name)
+            }}
+            className="key-button popup-item-key"
+          >
+            <span>{name}</span>
+          </button>
+        ))}
+        {
```
Comment (njm222) [inline src/ui/Keys.tsx:37]:
```suggestion
```

File: src/ui/Keys.tsx [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+          <button
+            className="add-button popup-item-key hovered-item"
+            onClick={() => {
+              onAdd(action)
+            }}
+          >
+            <span>+</span>
+          </button>
+        }
```
Comment (njm222) [inline src/ui/Keys.tsx:46]:
```suggestion
```

**Review by njm222 — APPROVED** [review]
LGTM! can you make the final change & resolve the merge conflict? and then we can merge it in! 🚀

File: src/ui/Keys.tsx (hunk 14 above)
Comment (littlemyx) [reply src/ui/Keys.tsx]: done!

**Comment by bjornstar** [comment]
Can you resolve the merge conflicts?

**Review by njm222 — APPROVED** [review]

**Comment by njm222** [comment]
@littlemyx  Can you resolve the CI checks as well?

**Comment by njm222** [comment]
@littlemyx it is still failing, can you make sure it is passing?

**Comment by littlemyx** [comment]
> @littlemyx it is still failing, can you make sure it is passing?

done! sorry for this, first-time-contributing errors 😞

**Comment by bjornstar** [comment]
We are unable to merge this branch due to conflicts, can you rebase against the main branch?
